pkg/test/dualstack/azure_dualstack_test.go

```diff
@@ -0,0 +1,581 @@
+			t.Logf("%q not healthy", pod.Name)
+		}
+
+		allHealthy = allHealthy && podHealthy
+	}
+
+	return allHealthy
+}
+
+func checkNodeReadiness(t *testing.T, userClient *kubernetes.Clientset) (string, error) {
+	expectedNodes := 1
+	var nodeIP string
+
+	err := wait.Poll(30*time.Second, 15*time.Minute, func() (bool, error) {
+		nodes, err := userClient.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+		if err != nil {
+			t.Logf("failed to get nodes list: %s", err)
+			return false, nil
+		}
+		if len(nodes.Items) != expectedNodes {
+			t.Logf("node count: %d, expected: %d", len(nodes.Items), expectedNodes)
+			return false, nil
+		}
+
+		readyNodeCount := 0
+
+		for _, node := range nodes.Items {
+			for _, c := range node.Status.Conditions {
+				if c.Type == corev1.NodeReady {
+					readyNodeCount++
+				}
+			}
+		}
+
+		if readyNodeCount != expectedNodes {
+			t.Logf("%d out of %d nodes are ready", readyNodeCount, expectedNodes)
+			return false, nil
+		}
+
+		for _, addr := range nodes.Items[0].Status.Addresses {
+			if addr.Type == corev1.NodeExternalIP {
+				nodeIP = addr.Address
+				break
+			}
+		}
+		return true, nil
+	})
+
+	return nodeIP, err
+}
+
+type cloudProvider interface {
+	NodeSpec() models.NodeCloudSpec
+	CloudSpec() models.CloudSpec
+}
+
+type azure struct{}
+
+var _ cloudProvider = azure{}
+
+func (a azure) NodeSpec() models.NodeCloudSpec {
+	return models.NodeCloudSpec{
+		Azure: &models.AzureNodeSpec{
+			AssignAvailabilitySet: true,
+			AssignPublicIP:        true,
+			DataDiskSize:          int32(30),
+			OSDiskSize:            70,
+			Size:                  pointer.String("Standard_B2s"),
+		},
+	}
+}
+
+func (a azure) CloudSpec() models.CloudSpec {
+	return models.CloudSpec{
+		DatacenterName: "azure-westeurope",
+		Azure: &models.AzureCloudSpec{
+			ClientID:        os.Getenv("AZURE_CLIENT_ID"),
+			ClientSecret:    os.Getenv("AZURE_CLIENT_SECRET"),
+			SubscriptionID:  os.Getenv("AZURE_SUBSCRIPTION_ID"),
+			TenantID:        os.Getenv("AZURE_TENANT_ID"),
+			LoadBalancerSKU: "standard",
+		},
+	}
+}
+
+func defaultClusterRequest() createClusterRequest {
+	clusterSpec := models.CreateClusterSpec{}
+	clusterSpec.Cluster = &models.Cluster{
+		Type: "kubernetes",
+		Name: "test-default-azure-cluster",
+		Spec: &models.ClusterSpec{
+			Cloud: &models.CloudSpec{
+				DatacenterName: "azure-westeurope",
+				Azure: &models.AzureCloudSpec{
+					ClientID:        os.Getenv("AZURE_CLIENT_ID"),
+					ClientSecret:    os.Getenv("AZURE_CLIENT_SECRET"),
+					SubscriptionID:  os.Getenv("AZURE_SUBSCRIPTION_ID"),
+					TenantID:        os.Getenv("AZURE_TENANT_ID"),
+					LoadBalancerSKU: "standard",
+				},
+			},
+			CniPlugin: &models.CNIPluginSettings{
+				Version: "v1.11",
+				Type:    "cilium",
+			},
+			ClusterNetwork: &models.ClusterNetworkingConfig{
+				NodeCIDRMaskSizeIPV4: 24,
+				// {"cidrBlocks":["172.25.0.0/16","fd00::/104"]},
+				// "services":{"cidrBlocks":["10.240.16.0/20","fd03::/120"]}
```

Comment:
the comment does not seem to have any special meaning, I guess you could remove it